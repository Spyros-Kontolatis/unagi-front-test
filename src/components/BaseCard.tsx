import React, { useState } from 'react';
import {
  CardWrapper,
  CardAction,
  CardContentSkeleton,
  CardSubtitle,
  CardTitle,
} from './BaseCard.style';
import type { BaseCardProps, Size, CardElementStyles } from '../types';

const configureStyleBasedOnSize = (size: Size): CardElementStyles => {
  switch (size) {
    case 'xs':
      return {
        width: '240px',
        height: '300px',
        titleFont: '18px',
        subtitleFont: '14px',
        padding: '16px 8px',
      };
    case 'sm':
      return {
        width: '300px',
        height: '375px',
        titleFont: '24px',
        subtitleFont: '16px',
        padding: '16px 8px',
      };
    case 'md':
      return {
        width: '400px',
        height: '500px',
        titleFont: '32px',
        subtitleFont: '18px',
        padding: '24px 8px',
      };
    case 'lg':
      return {
        width: '500px',
        height: '640px',
        titleFont: '36px',
        subtitleFont: '24px',
        padding: '32px 16px',
      };
  }
};

export default ({
  imageUrl,
  title,
  subtitle,
  size,
  action,
  loading,
  lazyLoading,
  onImgClick,
}: BaseCardProps) => {
  const [imgLoading, setImgLoading] = useState(lazyLoading);
  const [imgSrc, setImgSrc] = useState(imageUrl);
  const styledElements = configureStyleBasedOnSize(size ?? 'sm');
  if (loading)
    return (
      <CardWrapper
        width={styledElements.width}
        height={styledElements.height}
        $padding={styledElements.padding}
      >
        <CardContentSkeleton height="70%" width="70%" />
        <CardContentSkeleton
          height={styledElements.titleFont}
          width="50%"
          $cardmargintop="24px"
        />
        <CardContentSkeleton height={styledElements.subtitleFont} width="50%" />
        <CardContentSkeleton
          height="50px"
          width="50%"
          $cardborderradius="24px"
        />
      </CardWrapper>
    );
  else
    return (
      <CardWrapper
        width={styledElements.width}
        height={styledElements.height}
        $padding={styledElements.padding}
      >
        {imageUrl ? (
          <img
            loading={lazyLoading ? 'lazy' : 'eager'}
            height="70%"
            width="70%"
            src={imgSrc}
            onLoad={() => setImgLoading(false)}
            onError={() =>
              setImgSrc(
                'https://images.fotmob.com/image_resources/playerimages/26166.png',
              )
            }
            onClick={onImgClick}
            style={{ cursor: onImgClick ? 'pointer' : 'auto' }}
          />
        ) : (
          <CardContentSkeleton height="58%" width="60%" />
        )}
        {imgLoading && (
          <CardContentSkeleton
            style={{ position: 'absolute' }}
            height="58%"
            width="60%"
          />
        )}

        <CardTitle $cardfontsize={styledElements.titleFont}>{title}</CardTitle>
        {subtitle && (
          <CardSubtitle $cardfontsize={styledElements.subtitleFont}>
            {subtitle}
          </CardSubtitle>
        )}
        {action && (
          <CardAction onClick={action.handler}>{action.label}</CardAction>
        )}
      </CardWrapper>
    );
};
