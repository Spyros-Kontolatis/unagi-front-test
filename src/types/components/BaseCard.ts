import type { Size } from '../global';

interface BaseCardAction {
  label: string;
  handler: () => void;
}

export interface BaseCardProps {
  imageUrl: string;
  title: string;
  subtitle?: string;
  size?: Size;
  action?: BaseCardAction;
  loading?: boolean;
  lazyLoading?: boolean;
}

export interface CardElementStyles {
  width: string | number;
  height: string | number;
  titleFont: string;
  subtitleFont: string;
  padding: string;
}
