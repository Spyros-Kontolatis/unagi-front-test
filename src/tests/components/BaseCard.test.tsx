import BaseCard from '../../components/BaseCard';
import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';

describe('BaseCard', () => {
  test('renders the component', () => {
    const { baseElement } = render(
      <BaseCard
        imageUrl="test"
        title="Test"
        subtitle="Test"
        size="sm"
        action={{
          label: 'Action',
          handler: jest.fn(),
        }}
        loading={false}
        lazyLoading={true}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });
  test('renders the component with loading', () => {
    const { baseElement } = render(
      <BaseCard
        imageUrl="test"
        title="Test"
        subtitle="Test"
        size="sm"
        action={{
          label: 'Action',
          handler: jest.fn(),
        }}
        loading={true}
        lazyLoading={true}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  test('calls action handler', async () => {
    const action = {
      label: 'Action',
      handler: jest.fn(),
    };
    render(
      <BaseCard
        imageUrl="test"
        title="Test"
        subtitle="Test"
        size="sm"
        action={action}
        loading={false}
        lazyLoading={true}
      />,
    );

    const actionButton = screen.getByText('Action');
    expect(actionButton).toBeTruthy();
    await fireEvent.click(actionButton);
    expect(action.handler).toHaveBeenCalledTimes(1);
  });
});
