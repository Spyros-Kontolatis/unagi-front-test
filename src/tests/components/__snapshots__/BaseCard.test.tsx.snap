// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseCard renders the component 1`] = `
<body>
  <div>
    <div
      class="sc-aYaIB cjAnBo"
      height="375px"
      padding="16px 8px"
      width="300px"
    >
      <img
        height="70%"
        loading="lazy"
        src="test"
        width="70%"
      />
      <div
        class="sc-dcJtft heCfXP"
        height="58%"
        style="position: absolute;"
        width="60%"
      />
      <div
        cardfontsize="24px"
        class="sc-gEvDqW gOzUcR"
      >
        Test
      </div>
      <div
        cardfontsize="16px"
        class="sc-eqUzNf jGOZXw"
      >
        Test
      </div>
      <button
        class="sc-fqkwJk fmQuZg"
      >
        Action
      </button>
    </div>
  </div>
</body>
`;

exports[`BaseCard renders the component with loading 1`] = `
<body>
  <div>
    <div
      class="sc-aYaIB cjAnBo"
      height="375px"
      padding="16px 8px"
      width="300px"
    >
      <div
        class="sc-dcJtft dwNegU"
        height="70%"
        width="70%"
      />
      <div
        cardmargintop="24px"
        class="sc-dcJtft jPrLlQ"
        height="24px"
        width="50%"
      />
      <div
        class="sc-dcJtft idqYND"
        height="16px"
        width="50%"
      />
      <div
        cardborderradius="24px"
        class="sc-dcJtft iyFqJj"
        height="50px"
        width="50%"
      />
    </div>
  </div>
</body>
`;
